Rework Svucrg extension to use a 3-bit PTE implementation. (#737)
riscv/riscv-cheri: src/cheri/cheri-pte-ext.adoc

NOTE: The {ctag} bit of the stored capability is checked _after_ it is potentially
cleared due to lack of <<c_perm>>.

If the CW bit is clear, the "no capability state", then the existing rules from <<section_priv_cheri_vmem>> are followed:

* When a capability load or AMO instruction is executed, the implementation
 clears the {ctag} bit of the capability read from the virtual page.
* When a capability store or AMO instruction is executed and the {ctag} bit
 of the capability being written is set, the implementation raises a
 {cheri_excep_name_pte_st}.

When the CW bit is set, the "capability state", then the behaviour of
capability loads is controlled by CRG and the behaviour of capability stores

> is controlled by CD.

Reviewer: Gating CRG behavior on CW means we would be in trouble if ever we wanted PTEs that grant W and not CW for (aliased/shared) pages that hold caps (such as those that have a W and CW PTE elsewhere).  I don't know that we want such a thing, but it'd be good to (informatively) `[NOTE]` that this is a constraint on software behavior.

Author: Why is this a problem though? I  think that you can have a {W, ~CW} alias of a {W, CW} page, because in Svucrg CW gates both capability reads and writes, so you can't ever load a valid capability through the ~CW alias anyway. I think CW is a somewhat misleading name for the bit in this model, perhaps simply calling it `C` bit might be better and mirrors what we do in the capability HW permission bits with the "C-permission bit".

When CW is set, the CRG bit indicates the current generation of the virtual memory page with
regards to the ongoing capability revocation cycle. Two schemes are permitted:

* A {cheri_excep_name_pte_ld} is raised when a capability load or AMO instruction is executed

>  with:
> ** <<c_perm>> granted, and
> ** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG, and
> ** PTE.U is set.

Reviewer: Purely grammatical suggestion:
```suggestion
 and:
** the authorizing capability grants <<c_perm>>,
** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG, and
** the virtual page's U bit is set.
```

Reviewer 2: "the authorizing capability grants" is a phrase we use elsewhere, so better to use it

* A {cheri_excep_name_pte_ld} is raised when a capability load or AMO instruction is executed

>  with:
> ** <<c_perm>> granted, and
> ** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG, and
> ** PTE.U is set, and
> ** the capability read from memory has its {ctag} set.

Reviewer: ```suggestion
 wherein:
** the authorizing capability grants <<c_perm>>,
** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG,
** the virtual page's U bit is set, and
** the capability read from memory has its {ctag} set.
```

of the capability being written is set and the CD bit is clear,
two schemes are permitted (also see <<section_hardware_pte_updates>>):

* The same behavior as when CW is clear, allowing software interpretation
 of this state.
* The implementation sets the CD bit. The PTE update must be
 atomic with respect to other accesses to the PTE, and must atomically check
 that the PTE is valid and grants sufficient permissions. Updates to the CD bit
 must be exact (i.e., not speculative), and observed in program order by the
 local hart. Furthermore, the PTE update must appear in the global memory order
 no later than the explicit memory access, or any subsequent explicit memory
 access to that virtual page by the local hart. The ordering on loads and
 stores provided by FENCE instructions and the acquire/release bits on atomic
 instructions also orders the PTE updates associated with those loads and
 stores as observed by remote harts.
 The PTE update is not required to be atomic with respect to the explicit memory
 access that caused the update, and the sequence is interruptible. However, the
 hart must not perform explicit memory accesses before the PTE update is globally

>  visible.

Reviewer 2: do we really need all this text? isn't it already specified by Svadu, so we can just use tha?

Author: Dropped and explicitly referenced Svadu.

NOTE: _{cheri_priv_crg_ext}_ requires that when CW, CD and CRG are all clear,
the implementation clears loaded tags and raises {cheri_excep_name_pte_st} when
the stored {ctag} is set. Other bit combinations when CW=0 are reserved

> for future extensions.

Reviewer 2: the table doesn't show that - it shows PTE.CW=0 and PTE.CD=0
I think you need a row for PTE.CW=0 and PTE.CD=0 separately and if you hit the reserved case what happens? what exception type?

NOTE: The {ctag} bit of the stored capability is checked _after_ it is potentially
cleared due to lack of <<c_perm>>.

If the CW bit is clear, the "no capability state", then the existing rules from <<section_priv_cheri_vmem>> are followed:

* When a capability load or AMO instruction is executed, the implementation
 clears the {ctag} bit of the capability read from the virtual page.
* When a capability store or AMO instruction is executed and the {ctag} bit
 of the capability being written is set, the implementation raises a
 {cheri_excep_name_pte_st}.

> When the CW bit is set, the "capability state", then the behaviour of
> capability loads is controlled by CRG and the behaviour of capability stores

Reviewer 2: ```suggestion
When the CW bit is set, the "capability state", then the behavior of
capability loads is controlled by CRG and the behavior of capability stores
```

is controlled by CD.

When CW is set, the CRG bit indicates the current generation of the virtual memory page with
regards to the ongoing capability revocation cycle. Two schemes are permitted:

* A {cheri_excep_name_pte_ld} is raised when a capability load or AMO instruction is executed
 and:
** the authorizing capability grants <<c_perm>>,
** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG, and
** the virtual page's PTE.U is set.
* A {cheri_excep_name_pte_ld} is raised when a capability load or AMO instruction is executed

>  wherein:

Reviewer 2: ```suggestion
 and:
```

for consistency with the first scheme above

Reviewer 2: in fact I think it's better not to list the first 3 points twice, and just add the optional check to the end of the first list.

** the authorizing capability grants <<c_perm>>,
** the virtual page's CRG bit does not equal <<sstatusreg_pte,sstatus>>.UCRG,
** the virtual page's PTE.U is set, and
** the capability read from memory has its {ctag} set.

The second option is enabled by the _{cheri_priv_crg_load_tag_ext}_ extension.

When CW is set, the CD bit indicates that a capability was stored to the
virtual page since the last time the CD bit was cleared.
When a capability store or AMO instruction is executed, the {ctag} bit
of the capability being written is set and the CD bit is clear,
two schemes are permitted (also see <<section_hardware_pte_updates>>):

* The same behavior as when CW is clear, allowing software interpretation
 of this state.
* The implementation sets the CD bit. The PTE update behaves in the same way as
 the D bit update described by the _Svadu_ extesion.

When CW, CD and CRG are all clear, the implementation is required to clear

> loaded tags and raise {cheri_excp_name_pte_st} when the stored {ctag} is set.

Reviewer 2: ```suggestion
loaded tags and raise {cheri_excep_name_pte_st} when the stored {ctag} is set.
```

> NOTE: Other CD and CRG combinations when CW=0 are reserved for future extension.
> The reserved PTE states behave as the CW=0 CD=0 CRG=0, unless a future extension

Reviewer 2: ```suggestion
NOTE: Other CD and CRG combinations when CW=0 are reserved for future extensions.
The reserved PTE states behave as the CW=0, CD=0 and CRG=0 state, unless a future extension
```
